File: #34_Search_for_a_Range/Solution.py
class Solution:
	# @param {integer[]} nums
	# @param {integer} target
	# @return {integer[]}
	def searchRange(self, nums, target):
		if not nums:
			return [-1, -1]
		i, j, mid = 0, len(nums) - 1, 0
		while i <= j:
			mid = i + (j - i) / 2
			if nums[mid] == target:
				break
			elif nums[mid] > target:
				j = mid - 1
			else:
				i = mid + 1
		if i > j:
			return [-1, -1]
		return [self.searchLeft(nums, target, i, mid), self.searchRight(nums, target, mid, j)]

	# hi always remains possible result
	def searchLeft(self, nums, target, lo, hi):
		if lo == hi:
			return lo
		mid = lo + (hi - lo) / 2
		if nums[mid] == target:
			return self.searchLeft(nums, target, lo, mid)
		return self.searchLeft(nums, target, mid + 1, hi)

	# lo always remains possible result
	def searchRight(self, nums, target, lo, hi):
		if lo == hi:
			return lo
		mid = lo + (hi - lo) / 2
		if mid == lo: # special case for right search,
			mid += 1
		if nums[mid] == target:
			return self.searchRight(nums, target, mid, hi)
		return self.searchRight(nums, target, lo, mid - 1)

sol = Solution()
t = [5, 7, 7, 8, 8, 10]
print sol.searchRange(t, 8)

File: #34_Search_for_a_Range/Solution.java
public class Solution {
	static int left, right;
    public static int[] searchRange(int[] nums, int target) {
    	/* find left bound */
    	left = -1; right = -1;
    	searchLeft(nums, target, 0, nums.length - 1);
    	searchRight(nums, target, 0, nums.length - 1);
    	// System.out.println("Left: " + left + ", Right: " + right);
    	if (left == right) {
    		if (left < 0 || left >= nums.length) return new int[]{-1, -1};
    		if (nums[left] != target) return new int[]{-1, -1};
    	}
    	if (left > right) return new int[]{-1, -1};
        int[] res = new int[2];
        res[0] = left; res[1] = right;
        return res;
    }

    private static void searchLeft(int[] nums, int target, int lo, int hi) {
    	if (lo > hi) return;
    	int mid = (lo + hi) / 2;
    	if (nums[mid] > target) {
    		left = mid - 1;
    		searchLeft(nums, target, lo, left);
    	}
    	else if (nums[mid] < target) {
    		left = mid + 1;
    		searchLeft(nums, target, left, hi);
    	}
    	else {
    		left = mid;
    		searchLeft(nums, target, lo, left - 1);
    	}
    }

    private static void searchRight(int[] nums, int target, int lo, int hi) {
    	if (lo > hi) return;
    	int mid = (lo + hi) / 2;
    	if (nums[mid] > target) {
    		right = mid - 1;
    		searchRight(nums, target, lo, right);
    	}
    	else if (nums[mid] < target) {
    		right = mid + 1;
    		searchRight(nums, target, right, hi);
    	}
    	else {
    		right = mid;
    		searchRight(nums, target, right + 1, hi);
    	}
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[] s = {2, 2};
    	int[] res = sol.searchRange(s, 3);
    	System.out.println(res[0] + ", " + res[1]);
    }
}